src/paragraphcommand.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module paragraph/paragraphcommand
+ */
+
+import Command from '@ckeditor/ckeditor5-core/src/command/command';
+
+/**
+ * The paragraph command.
+ *
+ * @extends module:core/command/command~Command
+ */
+export default class ParagraphCommand extends Command {
+	/**
+	 * Creates an instance of the command.
+	 *
+	 * @param {module:core/editor/editor~Editor} editor Editor instance.
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		const t = editor.t;
+
+		/**
+		 * Value of the command, indicating whether it is applied in the context
+		 * of current {@link module:engine/model/document~Document#selection selection}.
+		 *
+		 * @readonly
+		 * @observable
+		 * @member {Boolean}
+		 */
+		this.set( 'value', false );
+
+		/**
+		 * User-readable title of the command, for use in the UI.
+		 *
+		 * @readonly
+		 * @member {String}
+		 */
+		this.title = t( 'Paragraph' );
+
+		// Update current value each time changes are done on document.
+		this.listenTo( editor.document, 'changesDone', () => this._updateValue() );
+	}
+
+	/**
+	 * Executes command.
+	 *
+	 * @protected
+	 * @param {Object} [options] Options for executed command.
+	 * @param {module:engine/model/batch~Batch} [options.batch] Batch to collect all the change steps.
+	 * New batch will be created if this option is not set.
+	 * @param {module:engine/model/element~Element} [options.element] Element the command should be applied to.
```

Comment:
Hm... I have mixed feelings about this implementation.

You're complicating the paragraph command because its expected, straightforward implementation didn't fit some use case. These are separate packages so we should think about them as about independent features. The change you did is confusing in context of the paragraph package. Apparently:

* this is not a use case for the paragraph command,
* or we need to find a more general implementation.

The first option would be simple – the heading command would handle this. It's applying the heading so it can remove it too.

OTOH, perhaps the logic of turning blocks into paragraphs should be enclosed in the paragraph package. Someone could, for instance, strip some attributes upon its execution. However, this is not ideal because a similar operation would be usually expected on the opposite action (applying a heading) too and this is a separate command still.

So, you have two choices:

* handle this in the heading command,
* allow passing a custom selection to the heading command (that's the more general implementation).

I'm fine with both, but the former seems to be simpler. It's hard for me to tell now which option is more future-safe.

BTW, I was actually very surprised that the heading command didn't clean up the entire selection. I remember we discussed this in the past and I can see that the command mimics CKE4's behaviour. Perhaps keeping it consistent with CKE4 makes a lot of sense, but for me it could also clean up the entire selection.